ts/DataGrid/DataGrid2/Virtualization/RowsVirtualizer.ts

```diff
@@ -0,0 +1,271 @@
+        const rowsPerPage = Math.ceil(
+            vp.tbodyElement.offsetHeight / this.defaultRowHeight
+        );
+
+        const rows = vp.rows;
+
+        if (!rows.length) {
+            const last = new DataGridRow(vp, vp.dataTable.getRowCount() - 1);
+            last.render(vp);
+            rows.push(last);
+            vp.tbodyElement.appendChild(last.htmlElement);
+        }
+
+        const from = Math.max(rowCursor - buffer, 0);
+        const to = Math.min(
+            rowCursor + rowsPerPage + buffer,
+            rows[rows.length - 1].index - 1
+        );
+
+        const alwaysLastRow = rows.pop();
+
+        for (let i = 0, iEnd = rows.length; i < iEnd; ++i) {
+            rows[i].destroy();
+        }
+        rows.length = 0;
+
+        for (let i = from; i <= to; ++i) {
+            const newRow = new DataGridRow(vp, i);
+            newRow.render(vp);
+            vp.tbodyElement.insertBefore(
+                newRow.htmlElement,
+                vp.tbodyElement.lastChild
+            );
+
+            rows.push(newRow);
+        }
+
+        if (alwaysLastRow) {
+            rows.push(alwaysLastRow);
+        }
+
+        const bof = buffer - (rowCursor - buffer < 0 ? buffer - rowCursor : 0);
+        this.topRowInitialHeight = vp.rows[bof].htmlElement.clientHeight;
+    }
+
+    /**
+     * Adjusts the heights of the rows based on the current scroll position.
+     * It handles the possibility of the rows having different heights than
+     * the default height.
+     */
+    public adjustRowHeights(): void {
+        const { rows, tbodyElement } = this.viewport;
+        const rowsLn = rows.length;
+
+        let translateBuffer = rows[0].getDefaultTopOffset();
+
+        for (let i = 0; i < rowsLn; ++i) {
+            const row = rows[i];
+            const cursor = this.rowCursor;
+
+            if (row.index > cursor) {
+                break;
+            }
+
+            const element = row.htmlElement;
+            const defaultH = this.defaultRowHeight;
+            const borderH = element.offsetHeight - element.clientHeight;
+
+            if (row.index < cursor) {
+                row.htmlElement.style.height = defaultH + borderH + 'px';
+            } else if (
+                row.getCurrentHeight() > defaultH &&
+                this.topRowInitialHeight
+            ) {
+                const ratio = tbodyElement.scrollTop / defaultH - cursor;
+                const diff = this.topRowInitialHeight - defaultH;
+
+                row.htmlElement.style.height =
+                    this.topRowInitialHeight - diff * ratio + 'px';
+            }
+        }
+
+        for (let i = 1, iEnd = rowsLn - 1; i < iEnd; ++i) {
+            translateBuffer += rows[i - 1].getCurrentHeight();
+            rows[i].htmlElement.style.transform =
+                `translateY(${translateBuffer}px)`;
+        }
+
+        if (rows[rowsLn - 2].index + 1 === rows[rowsLn - 1].index) {
```

Comment:
Is this "the last object"? If yes, maybe it would be good to add props "isLast" or "isFirst" when initializing rows. 

Its pretty the same like in HC we use flags for labels.

Reply:
Is it necessary? The first one is always the one with id 0, and the last one with ID length - 1. Here we check whether the penultimate one is rendered, if so, it should change the offset for the last one as well.